olive/passes/pytorch/pipeline_parallel.py

```diff
@@ -0,0 +1,153 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+# Automatically distribute PyTorch model using Pipeline Parallelism
+# --------------------------------------------------------------------------
+
+import logging
+import multiprocessing
+from abc import abstractmethod
+from pathlib import Path
+from typing import Any, Callable, Dict
+
+import torch
+
+from olive.common.pydantic_v1 import validator
+from olive.hardware.accelerator import AcceleratorSpec, Device
+from olive.model import DistributedPyTorchModelHandler, PyTorchModelHandler
+from olive.passes import Pass
+from olive.passes.olive_pass import PassConfigParam
+
+logger = logging.getLogger(__name__)
+
+
+class PipelineParallel:
+    def __init__(self, rank: int, world_size: int):
+        self.rank = rank
+        self.world_size = world_size
+
+    @abstractmethod
+    def replace_layers(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def restore_layers(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def split_layers(self, model: torch.nn.Module):
+        raise NotImplementedError
+
+
+class PyTorchPipelineParallel(Pass):
+    @staticmethod
+    def _default_config(accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        # Note : The default value for nstages should be the no of gpus
+        # (AcceleratorSpec.Device == GPU) in the target OliveSystem
+        return {
+            "nstages": PassConfigParam(
+                type_=int,
+                default=2,
+                required=True,
+                description="Number of stages to distribute the model for. Must be greater than 1.",
+            ),
+            "parallel_jobs": PassConfigParam(
+                type_=int,
+                default=multiprocessing.cpu_count(),
+                required=False,
+                description="Number of parallel jobs. Defaulted to number of CPUs. Set it to 0 to disable.",
+            ),
+        }
+
+    @staticmethod
+    def _validate_nstages(v):
+        if int(v) < 2:
+            raise ValueError("nstages should be >= 2")
+
+        return v
+
+    @staticmethod
+    def _validators() -> Dict[str, Callable]:
```

Comment:
please rebase from main, the signature is changed to classmethod instead of staticmethod